# Day 25 — NumPy Fundamentals
 
**Week 5: NumPy — Numerical Computing Foundation**
 
Topics covered: `ndarray` creation, `dtype`, `shape`, `ndim`, `size`, array from lists/ranges, `zeros`, `ones`, `full`, `eye`, `arange`, `linspace`, `reshape`, basic indexing, `astype`, comparing arrays to Python lists.
 
---

### Q1. Your First NumPy Array
Import NumPy as `np`. Create a 1D array from the Python list `[10, 20, 30, 40, 50]`. Print the array, its `dtype`, `shape`, `ndim`, and `size`.

In [1]:
# Q1: Create a 1D NumPy array and inspect its properties

import numpy as np

arr = np.array([10, 20, 30, 40, 50])  # create array from Python list

print("Array:  ", arr)
print("dtype:  ", arr.dtype)    # data type of elements (e.g. int64)
print("shape:  ", arr.shape)    # tuple showing dimensions (rows, cols, ...)
print("ndim:   ", arr.ndim)     # number of dimensions
print("size:   ", arr.size)     # total number of elements

Array:   [10 20 30 40 50]
dtype:   int64
shape:   (5,)
ndim:    1
size:    5


### How it works

- `np.array()` converts a Python list into a NumPy `ndarray` (n-dimensional array).
- `dtype` tells you the type of each element — NumPy auto-detected `int64` here.
- `shape` returns a tuple: `(5,)` means 1 row with 5 elements (1D array).
- `ndim` is the number of dimensions — `1` for a flat list-like array.
- `size` is the total count of elements — same as `len()` for 1D arrays.

**Key takeaway:** Every NumPy array carries metadata (`dtype`, `shape`, `ndim`, `size`) that tells you exactly what kind of data it holds and how it is structured.

### Q2. Array from Nested List (2D)
Create a 2D NumPy array from the nested list `[[1, 2, 3], [4, 5, 6], [7, 8, 9]]`. Print its `shape`, `ndim`, number of rows, number of columns, and total element count.

In [2]:
# Q2: Create a 2D array from a nested list and inspect its structure

import numpy as np

arr = np.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])  # nested list → 2D array

print("Array:\n", arr)
print("shape:   ", arr.shape)       # (rows, cols)
print("ndim:    ", arr.ndim)        # 2 for a matrix
print("rows:    ", arr.shape[0])    # first element of shape tuple
print("cols:    ", arr.shape[1])    # second element of shape tuple
print("size:    ", arr.size)        # total elements = rows × cols

Array:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
shape:    (3, 3)
ndim:     2
rows:     3
cols:     3
size:     9


### How it works

- A nested list (list of lists) gives NumPy the data for a 2D array — think of it as a table with rows and columns.
- `shape` returns `(3, 3)`: 3 rows and 3 columns.
- `shape[0]` is the row count; `shape[1]` is the column count.
- `ndim` is `2` — one dimension for rows, one for columns.
- `size` equals `rows × cols` = 3 × 3 = 9.

**Key takeaway:** For any 2D array, `shape` gives `(rows, cols)` and you can unpack them individually with `shape[0]` and `shape[1]`.

### Q3. NumPy vs Python List — Speed Comparison
Create a Python list and a NumPy array both containing integers from 0 to 999,999. Using the `time` module, measure and compare the time to square every element (use a list comprehension for the list, and direct `** 2` for the array). Print the speedup factor.

In [3]:
# Q3: Compare speed of squaring elements — Python list vs NumPy array

import numpy as np
import time

# --- Python list ---
py_list = list(range(1_000_000))

start = time.time()
squared_list = [x ** 2 for x in py_list]   # list comprehension, element by element
list_time = time.time() - start

# --- NumPy array ---
np_arr = np.arange(1_000_000)

start = time.time()
squared_arr = np_arr ** 2                   # vectorized: operates on whole array at once
numpy_time = time.time() - start

# --- Results ---
speedup = list_time / numpy_time

print(f"Python list time : {list_time:.4f} seconds")
print(f"NumPy array time : {numpy_time:.4f} seconds")
print(f"NumPy is ~{speedup:.1f}x faster than Python list")

Python list time : 0.0641 seconds
NumPy array time : 0.0017 seconds
NumPy is ~37.9x faster than Python list


### How it works

- The Python list approach loops through 1,000,000 elements one by one using a list comprehension.
- NumPy's `** 2` is **vectorized** — it applies the operation to all elements in one low-level C call, with no Python loop.
- `time.time()` captures the clock before and after each operation; the difference is elapsed time.
- Dividing `list_time / numpy_time` gives the speedup factor — typically 10x–100x in favor of NumPy.
- This is the core reason NumPy exists: Python loops are slow; NumPy pushes the loop down to compiled C code.

**Key takeaway:** NumPy's vectorized operations avoid Python-level loops entirely, making numerical computation dramatically faster on large arrays.

### Q4. NumPy vs Python List — Memory Comparison
Using `sys.getsizeof` and `array.nbytes`, compare the memory used by a Python list of 10,000 floats vs a NumPy float64 array of the same values. Then compare again using `float32`. Print all three sizes in KB.

In [4]:
# Q4: Compare memory usage — Python list vs NumPy float64 vs float32

import numpy as np
import sys

data = [float(x) for x in range(10_000)]   # Python list of 10,000 floats

# --- Python list memory ---
list_bytes = sys.getsizeof(data)            # size of the list object itself (bytes)

# --- NumPy float64 (default) ---
arr_f64 = np.array(data, dtype=np.float64)
f64_bytes = arr_f64.nbytes                  # total bytes used by array elements

# --- NumPy float32 (half the precision) ---
arr_f32 = np.array(data, dtype=np.float32)
f32_bytes = arr_f32.nbytes

# --- Print sizes in KB ---
print(f"Python list (float)   : {list_bytes / 1024:.2f} KB")
print(f"NumPy float64         : {f64_bytes  / 1024:.2f} KB")
print(f"NumPy float32         : {f32_bytes  / 1024:.2f} KB")
print(f"\nfloat64 vs list       : NumPy uses {list_bytes / f64_bytes:.1f}x less memory")
print(f"float32 vs float64    : float32 uses {f64_bytes / f32_bytes:.1f}x less memory")

Python list (float)   : 83.18 KB
NumPy float64         : 78.12 KB
NumPy float32         : 39.06 KB

float64 vs list       : NumPy uses 1.1x less memory
float32 vs float64    : float32 uses 2.0x less memory


### How it works

- `sys.getsizeof(list)` returns the memory of the list's internal pointer array — each float object in a Python list is a separate heavyweight Python object stored elsewhere in memory, so the real usage is even higher.
- `array.nbytes` returns the exact bytes used by NumPy array elements: `num_elements × bytes_per_dtype`.
- `float64` uses 8 bytes per element → 10,000 × 8 = 80,000 bytes (~78 KB).
- `float32` uses 4 bytes per element → 10,000 × 4 = 40,000 bytes (~39 KB) — half the size.
- Choosing a smaller `dtype` like `float32` is a simple way to halve memory when full 64-bit precision is not needed.

**Key takeaway:** NumPy stores data in a compact, typed memory block (`nbytes`), while Python lists store pointers to individual objects — making NumPy far more memory-efficient for large numerical data.

### Q5. `np.zeros`, `np.ones`, `np.full`
Create the following without using Python loops:
- A 3×4 matrix of all zeros (`float64`)
- A 2×5 matrix of all ones (`int32`)
- A 4×4 matrix where every element is `7` (`float32`)
Print each with its dtype.
 

In [5]:
# Q5: Create pre-filled arrays using zeros, ones, and full

import numpy as np

# --- 3×4 matrix of zeros (float64) ---
zeros = np.zeros((3, 4), dtype=np.float64)

# --- 2×5 matrix of ones (int32) ---
ones = np.ones((2, 5), dtype=np.int32)

# --- 4×4 matrix filled with 7 (float32) ---
sevens = np.full((4, 4), fill_value=7, dtype=np.float32)

# --- Print each with dtype ---
print("zeros (float64):\n", zeros)
print("dtype:", zeros.dtype, "\n")

print("ones (int32):\n", ones)
print("dtype:", ones.dtype, "\n")

print("sevens (float32):\n", sevens)
print("dtype:", sevens.dtype)

zeros (float64):
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
dtype: float64 

ones (int32):
 [[1 1 1 1 1]
 [1 1 1 1 1]]
dtype: int32 

sevens (float32):
 [[7. 7. 7. 7.]
 [7. 7. 7. 7.]
 [7. 7. 7. 7.]
 [7. 7. 7. 7.]]
dtype: float32


### How it works

- All three functions take a **shape tuple** `(rows, cols)` as the first argument — no Python loop needed.
- `np.zeros()` fills every element with `0.0`; useful as a blank slate for calculations.
- `np.ones()` fills every element with `1`; commonly used for initializing masks or weight arrays.
- `np.full()` fills every element with any value you choose via `fill_value`.
- The `dtype` argument controls the number type — if omitted, `zeros` and `ones` default to `float64`, and `full` infers from the fill value.

**Key takeaway:** `zeros`, `ones`, and `full` are the standard way to create pre-filled arrays of any shape and dtype in one line — no loops, no list comprehensions.

### Q6. `np.eye` and `np.identity`
Create a 5×5 identity matrix using `np.eye`. Then use `np.eye` with a `k` offset to create a matrix with ones on the first superdiagonal (above main diagonal) and another with ones on the first subdiagonal. Print all three.

In [6]:
# Q6: Identity matrix and diagonal offsets using np.eye

import numpy as np

# --- 5×5 identity matrix (ones on main diagonal, k=0 by default) ---
identity = np.eye(5)

# --- Ones on first superdiagonal (one above main, k=+1) ---
superdiag = np.eye(5, k=1)

# --- Ones on first subdiagonal (one below main, k=-1) ---
subdiag = np.eye(5, k=-1)

print("Identity matrix (k=0):\n", identity, "\n")
print("Superdiagonal (k=+1):\n", superdiag, "\n")
print("Subdiagonal (k=-1):\n", subdiag)

Identity matrix (k=0):
 [[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]] 

Superdiagonal (k=+1):
 [[0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]] 

Subdiagonal (k=-1):
 [[0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]]


### How it works

- `np.eye(N)` creates an N×N matrix with `1.0` on the main diagonal and `0.0` everywhere else — this is the **identity matrix**.
- The optional `k` parameter shifts where the ones are placed:
  - `k=0` → main diagonal (default)
  - `k=+1` → one step above the main diagonal (superdiagonal)
  - `k=-1` → one step below the main diagonal (subdiagonal)
- Think of `k` as a vertical shift: positive moves the ones up, negative moves them down.
- `np.identity(N)` also creates an identity matrix but has no `k` parameter — `np.eye` is more flexible.

**Key takeaway:** `np.eye` with the `k` offset lets you place ones on any diagonal — positive `k` goes above the main diagonal, negative `k` goes below it.

### Q7. `np.arange`
Use `np.arange` to create:
- Integers from 0 to 19
- Even integers from 0 to 40
- A descending sequence from 10 to 1 (inclusive)
- A float sequence from 0.0 to 1.0 in steps of 0.1
Print all four arrays.

In [7]:
# Q7: Generate sequences using np.arange

import numpy as np

# --- Integers from 0 to 19 (stop is exclusive) ---
seq1 = np.arange(0, 20)

# --- Even integers from 0 to 40 (step=2) ---
seq2 = np.arange(0, 41, 2)          # stop=41 so that 40 is included

# --- Descending from 10 to 1 (step=-1) ---
seq3 = np.arange(10, 0, -1)         # stop=0 so that 1 is included

# --- Float sequence 0.0 to 1.0 in steps of 0.1 ---
seq4 = np.arange(0.0, 1.1, 0.1)     # stop=1.1 to include 1.0

print("0 to 19:          ", seq1)
print("Evens 0 to 40:    ", seq2)
print("Descending 10→1:  ", seq3)
print("Float 0.0 to 1.0: ", np.round(seq4, 1))   # round to avoid float drift

0 to 19:           [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
Evens 0 to 40:     [ 0  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40]
Descending 10→1:   [10  9  8  7  6  5  4  3  2  1]
Float 0.0 to 1.0:  [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


### How it works

- `np.arange(start, stop, step)` works like Python's `range()` but returns a NumPy array and supports float steps.
- The `stop` value is always **exclusive** — so to include `40`, you pass `41`; to include `1.0`, you pass `1.1`.
- A negative `step` makes a descending sequence — `start` must be greater than `stop` in that case.
- Float steps can cause tiny precision errors (e.g. `0.30000000000000004`), so `np.round()` is used to display cleanly — the underlying values are still correct.
- For float sequences where exact endpoint inclusion matters, `np.linspace` (next topic) is usually safer than `np.arange`.

**Key takeaway:** `np.arange(start, stop, step)` generates sequences like `range()` but with array output and float support — always remember `stop` is exclusive.

### Q8. `np.linspace`
Use `np.linspace` to create:
- 10 evenly spaced values between 0 and 1 (inclusive of both endpoints)
- 7 evenly spaced values between -π and π
- 5 evenly spaced values between 1 and 100 with `endpoint=False`
Explain in a comment the key difference between `linspace` and `arange`.

In [8]:
# Q8: Generate evenly spaced values using np.linspace
# Key difference:
#   np.arange(start, stop, step) — you control the STEP SIZE, stop is exclusive
#   np.linspace(start, stop, num) — you control the COUNT of points, stop is inclusive by default

import numpy as np

# --- 10 values from 0 to 1 (both endpoints included) ---
seq1 = np.linspace(0, 1, 10)

# --- 7 values from -π to π ---
seq2 = np.linspace(-np.pi, np.pi, 7)

# --- 5 values from 1 to 100, endpoint excluded ---
seq3 = np.linspace(1, 100, 5, endpoint=False)   # stop=100 is NOT included

print("10 values, 0 to 1:          ", np.round(seq1, 4))
print("7 values, -π to π:          ", np.round(seq2, 4))
print("5 values, 1 to 100 (no end):", np.round(seq3, 4))

10 values, 0 to 1:           [0.     0.1111 0.2222 0.3333 0.4444 0.5556 0.6667 0.7778 0.8889 1.    ]
7 values, -π to π:           [-3.1416 -2.0944 -1.0472  0.      1.0472  2.0944  3.1416]
5 values, 1 to 100 (no end): [ 1.  20.8 40.6 60.4 80.2]


### How it works

- `np.linspace(start, stop, num)` generates exactly `num` evenly spaced values — you control the **count**, not the step size.
- By default both endpoints are included; `endpoint=False` excludes the stop value (useful for periodic signals or circular data).
- NumPy provides `np.pi` as a built-in constant — no `import math` needed.
- The step size is calculated automatically as `(stop - start) / (num - 1)` when `endpoint=True`.
- Unlike `np.arange`, `linspace` never has float drift issues because it divides the interval mathematically rather than accumulating steps.

**Key takeaway:** Use `np.arange` when you know the step size; use `np.linspace` when you know how many points you need — `linspace` is safer and more precise for float ranges.

### Q9. `reshape`
Create a 1D array of integers from 1 to 24. Reshape it into:
- A 4×6 matrix
- A 2×3×4 3D array
- A 6×4 matrix
Print each shape. Then demonstrate that `reshape` returns a **view** (modifying one modifies the other) by changing an element.

In [9]:
# Q9: Reshape a 1D array into multiple shapes and prove reshape returns a view

import numpy as np

# --- Base 1D array: integers 1 to 24 ---
arr = np.arange(1, 25)               # shape: (24,)

# --- Reshape into 4×6 matrix ---
mat_4x6 = arr.reshape(4, 6)          # 4 rows, 6 cols  → 4×6 = 24 ✓

# --- Reshape into 2×3×4 3D array ---
arr_3d = arr.reshape(2, 3, 4)        # 2 blocks, 3 rows, 4 cols → 2×3×4 = 24 ✓

# --- Reshape into 6×4 matrix ---
mat_6x4 = arr.reshape(6, 4)          # 6 rows, 4 cols  → 6×4 = 24 ✓

print("Original shape :", arr.shape)
print("4×6 shape      :", mat_4x6.shape)
print("2×3×4 shape    :", arr_3d.shape)
print("6×4 shape      :", mat_6x4.shape)

# --- Prove reshape returns a VIEW, not a copy ---
print("\nBefore modification:")
print("  arr[0]       =", arr[0])
print("  mat_4x6[0,0] =", mat_4x6[0, 0])   # same element as arr[0]

arr[0] = 999                                 # modify the original

print("\nAfter setting arr[0] = 999:")
print("  arr[0]       =", arr[0])
print("  mat_4x6[0,0] =", mat_4x6[0, 0])   # also changed — it's a view!

Original shape : (24,)
4×6 shape      : (4, 6)
2×3×4 shape    : (2, 3, 4)
6×4 shape      : (6, 4)

Before modification:
  arr[0]       = 1
  mat_4x6[0,0] = 1

After setting arr[0] = 999:
  arr[0]       = 999
  mat_4x6[0,0] = 999


### How it works

- `reshape(rows, cols)` returns a new **view** of the same data — no copying, just a different lens over the same memory block.
- The total number of elements must stay the same: `4×6 = 2×3×4 = 6×4 = 24`.
- For 3D arrays, the shape tuple has three values: `(blocks, rows, cols)`.
- Because `mat_4x6` is a view of `arr`, modifying `arr[0]` instantly changes `mat_4x6[0, 0]` — they point to the same memory location.
- To get an independent copy instead of a view, use `.copy()`: `arr.reshape(4, 6).copy()`.

**Key takeaway:** `reshape` returns a view of the original data — fast and memory-efficient — but changes to either the original or the reshaped array will affect both.

### Q10. `np.random` — Random Array Creation
Using `np.random`:
- Create a 3×3 matrix of random floats between 0 and 1 (`np.random.rand`)
- Create a 4×4 matrix of random integers between 1 and 100 (`np.random.randint`)
- Create a 1D array of 8 values drawn from a standard normal distribution (`np.random.randn`)
- Set a seed with `np.random.seed(42)` and verify reproducibility by running the same call twice

In [10]:
# Q10: Generate random arrays using np.random

import numpy as np

# --- Seed for reproducibility ---
np.random.seed(42)                              # fix the random number generator state

# --- 3×3 matrix of floats in [0.0, 1.0) ---
rand_floats = np.random.rand(3, 3)             # uniform distribution between 0 and 1

# --- 4×4 matrix of integers between 1 and 100 ---
rand_ints = np.random.randint(1, 101, size=(4, 4))  # low=1, high=101 (exclusive)

# --- 1D array of 8 values from standard normal distribution ---
rand_normal = np.random.randn(8)               # mean=0, std=1 (bell curve)

print("rand (3×3 floats):\n", np.round(rand_floats, 4), "\n")
print("randint (4×4, 1–100):\n", rand_ints, "\n")
print("randn (8 normal values):\n", np.round(rand_normal, 4), "\n")

# --- Verify reproducibility: same seed → same output ---
print("Reproducibility check (seed=42):")
np.random.seed(42)
first_run  = np.random.rand(3)
np.random.seed(42)
second_run = np.random.rand(3)

print("  First run : ", np.round(first_run,  4))
print("  Second run: ", np.round(second_run, 4))
print("  Identical?", np.array_equal(first_run, second_run))

rand (3×3 floats):
 [[0.3745 0.9507 0.732 ]
 [0.5987 0.156  0.156 ]
 [0.0581 0.8662 0.6011]] 

randint (4×4, 1–100):
 [[24  3 22 53]
 [ 2 88 30 38]
 [ 2 64 60 21]
 [33 76 58 22]] 

randn (8 normal values):
 [-2.4391  0.6034 -0.251  -0.1639 -1.4763  1.487  -0.0245  0.3556] 

Reproducibility check (seed=42):
  First run :  [0.3745 0.9507 0.732 ]
  Second run:  [0.3745 0.9507 0.732 ]
  Identical? True


### How it works

- `np.random.rand(rows, cols)` generates floats uniformly in `[0.0, 1.0)` — note the upper bound is never reached.
- `np.random.randint(low, high, size)` generates integers in `[low, high)` — so pass `101` to include `100`.
- `np.random.randn(n)` draws from a **standard normal distribution** (mean = 0, std = 1) — most values fall between -3 and +3.
- `np.random.seed(42)` locks the random number generator to a fixed starting point — same seed always produces the same sequence of numbers.
- Setting a seed is essential in data science and ML so experiments can be reproduced exactly by others.

**Key takeaway:** Always set `np.random.seed()` at the top of any experiment involving randomness — it makes your results reproducible and shareable.

### Q11. `dtype` Conversion with `astype`
Create an array `[1.7, 2.3, 3.9, 4.1]` as `float64`. Convert it to `int32` (observe truncation, not rounding), then to `bool`, then to `complex`. Print each conversion result and its dtype.

In [13]:
# Q11: Convert array dtypes using astype and observe behavior

import numpy as np

# --- Base array as float64 ---
arr = np.array([1.7, 2.3, 3.9, 4.1], dtype=np.float64)

# --- Convert to int32 (truncates toward zero, does NOT round) ---
arr_int = arr.astype(np.int32)

# --- Convert to bool (0 → False, any non-zero → True) ---
arr_bool = arr_int.astype(np.bool_)

# --- Convert to complex (adds imaginary part 0j) ---
arr_complex = arr.astype(np.complex128)

print("Original float64: ", arr,         "| dtype:", arr.dtype)
print("→ int32:          ", arr_int,     "| dtype:", arr_int.dtype)
print("→ bool:           ", arr_bool,    "| dtype:", arr_bool.dtype)
print("→ complex128:     ", arr_complex, "| dtype:", arr_complex.dtype)

Original float64:  [1.7 2.3 3.9 4.1] | dtype: float64
→ int32:           [1 2 3 4] | dtype: int32
→ bool:            [ True  True  True  True] | dtype: bool
→ complex128:      [1.7+0.j 2.3+0.j 3.9+0.j 4.1+0.j] | dtype: complex128


### How it works

- `astype()` always returns a **new array** — the original is never modified.
- `float64 → int32` **truncates** the decimal part toward zero: `1.7` becomes `1`, `3.9` becomes `3` — no rounding occurs.
- `int32 → bool` follows the rule: `0` becomes `False`, any non-zero value becomes `True`.
- `float64 → complex128` keeps the original value as the real part and appends `+0j` as the imaginary part.
- Conversion can cause **data loss** (e.g. float → int drops decimals) — always be intentional about dtype changes.

**Key takeaway:** `astype()` converts dtypes but truncates rather than rounds when going from float to int — always check your values after conversion to avoid silent data loss.

### Q12. `np.zeros_like`, `np.ones_like`, `np.full_like`
Given an existing array, use `zeros_like`, `ones_like`, and `full_like` to create new arrays with the same shape and dtype. Demonstrate the usefulness of these when you need a same-shape placeholder without knowing the dimensions in advance.

In [14]:
# Q12: Create same-shape arrays using zeros_like, ones_like, full_like

import numpy as np

# --- Source array (imagine this came from a file or function — shape unknown in advance) ---
source = np.array([[3.5, 1.2, 4.8],
                   [0.7, 2.1, 9.3]], dtype=np.float32)

# --- Same shape and dtype, filled with zeros ---
like_zeros = np.zeros_like(source)

# --- Same shape and dtype, filled with ones ---
like_ones = np.ones_like(source)

# --- Same shape and dtype, filled with a custom value ---
like_full = np.full_like(source, fill_value=99.9)

print("Source array (float32):\n", source, "\n")
print("zeros_like:\n", like_zeros, "| dtype:", like_zeros.dtype, "\n")
print("ones_like:\n",  like_ones,  "| dtype:", like_ones.dtype,  "\n")
print("full_like:\n",  like_full,  "| dtype:", like_full.dtype,  "\n")

# --- Practical demo: create a result placeholder matching an unknown input ---
def process(data):
    result = np.zeros_like(data)        # safe placeholder — matches whatever shape arrives
    result = data * 2                   # example operation
    return result

output = process(source)
print("process() output shape:", output.shape, "| dtype:", output.dtype)

Source array (float32):
 [[3.5 1.2 4.8]
 [0.7 2.1 9.3]] 

zeros_like:
 [[0. 0. 0.]
 [0. 0. 0.]] | dtype: float32 

ones_like:
 [[1. 1. 1.]
 [1. 1. 1.]] | dtype: float32 

full_like:
 [[99.9 99.9 99.9]
 [99.9 99.9 99.9]] | dtype: float32 

process() output shape: (2, 3) | dtype: float32


### How it works

- `zeros_like`, `ones_like`, and `full_like` each take an **existing array** and produce a new array with the same `shape` and `dtype` — no need to hardcode dimensions.
- This is especially useful inside functions where the input shape is unknown until runtime.
- `full_like` inherits the source `dtype`, so `fill_value=99.9` stored in a `float32` array stays `float32` — no dtype mismatch.
- Think of these as "carbon copy containers" — same mold, different content.
- They are safer than `np.zeros((arr.shape[0], arr.shape[1]))` because you cannot accidentally mis-specify the shape.

**Key takeaway:** Use `_like` functions when you need a same-shape placeholder — they automatically match the source array's shape and dtype, making your code robust to inputs of any size.

### Q13. `np.empty`
Create a 3×3 array using `np.empty`. Print it and explain in a comment why the values are unpredictable, when `empty` is preferable to `zeros`, and what risk comes with using uninitialized memory.

In [15]:
# Q13: Create an uninitialized array with np.empty

import numpy as np

# --- Create a 3×3 empty array (float64 by default) ---
empty_arr = np.empty((3, 3))

# --- Also create one with explicit dtype ---
empty_int = np.empty((2, 4), dtype=np.int32)

print("np.empty (float64):\n", empty_arr, "\n")
print("np.empty (int32):\n", empty_int, "\n")

# --- Speed comparison: empty vs zeros ---
import time

N = 10_000_000

start = time.time()
_ = np.zeros(N)
zeros_time = time.time() - start

start = time.time()
_ = np.empty(N)
empty_time = time.time() - start

print(f"np.zeros time : {zeros_time:.5f} sec")
print(f"np.empty time : {empty_time:.5f} sec")
print(f"empty is ~{zeros_time / empty_time:.1f}x faster than zeros")

np.empty (float64):
 [[0.3745 0.9507 0.732 ]
 [0.5987 0.156  0.156 ]
 [0.0581 0.8662 0.6011]] 

np.empty (int32):
 [[0 0 0 0]
 [0 0 0 0]] 

np.zeros time : 0.00066 sec
np.empty time : 0.00012 sec
empty is ~5.5x faster than zeros


### How it works

- `np.empty()` allocates a block of memory **without initializing it** — the values are whatever bytes happened to be at that memory location (leftover data from previous operations).
- This is why the printed values look random or garbage-like — they are not random, just uninitialized.
- `empty` is faster than `zeros` or `ones` because it skips the extra step of writing fill values into every element.
- It is preferable to `zeros` when you know you will **overwrite every element immediately** — for example, filling results inside a loop — so initialization is wasted work.
- The risk: if you accidentally read an uninitialized element before writing to it, you get silent garbage data with no error or warning.

**Key takeaway:** Use `np.empty` only when you will fill every element yourself right away — it is faster than `zeros` but dangerous if any element is read before being written.

### Q14. `np.array` with `copy` parameter
Create an array `a`. Create `b = np.array(a, copy=False)` and `c = np.array(a, copy=True)`. Modify `a[0]`. Check whether `b` and `c` reflect the change. Explain the behavior.

In [17]:
# Q14: Understand copy=False vs copy=True in np.array()

import numpy as np

# --- Original array ---
a = np.array([10, 20, 30, 40, 50])

# --- copy=False: reuse memory if possible (may return a view) ---
b = np.array(a, copy=False)

# --- copy=True: always allocate new memory (independent copy) ---
c = np.array(a, copy=True)

print("Before modifying a:")
print("  a =", a)
print("  b =", b)
print("  c =", c)

# --- Modify the first element of a ---
a[0] = 999

print("\nAfter setting a[0] = 999:")
print("  a =", a)
print("  b =", b,  "← shares memory with a (copy=False)")
print("  c =", c,  "← independent copy, unchanged (copy=True)")

# --- Confirm with np.shares_memory ---
print("\nb shares memory with a:", np.shares_memory(a, b))
print("c shares memory with a:", np.shares_memory(a, c))

Before modifying a:
  a = [10 20 30 40 50]
  b = [10 20 30 40 50]
  c = [10 20 30 40 50]

After setting a[0] = 999:
  a = [999  20  30  40  50]
  b = [999  20  30  40  50] ← shares memory with a (copy=False)
  c = [10 20 30 40 50] ← independent copy, unchanged (copy=True)

b shares memory with a: True
c shares memory with a: False


### How it works

- `np.array(a, copy=False)` avoids allocating new memory when possible — `b` points to the same underlying data as `a`, so changes to `a` are visible in `b`.
- `np.array(a, copy=True)` always allocates a fresh block of memory — `c` is fully independent and never affected by changes to `a`.
- `np.shares_memory(x, y)` is the reliable way to confirm whether two arrays share the same memory block.
- `copy=False` does not *guarantee* a view — if the dtype or layout forces a conversion, NumPy may copy anyway. `copy=True` always guarantees a copy.
- Think of `copy=False` as "borrow the original notebook" and `copy=True` as "print a fresh photocopy."

**Key takeaway:** Use `copy=True` when you need to modify an array safely without affecting the source; use `copy=False` for memory efficiency when you only need to read the data.

### Q15. Array Attributes Deep Dive
Create a 3D array of shape `(2, 3, 4)`. Print and explain every attribute: `shape`, `ndim`, `size`, `dtype`, `itemsize`, `nbytes`, `strides`. Calculate manually what `nbytes` should be from `size * itemsize` and verify.

In [18]:
# Q15: Explore every attribute of a 3D NumPy array

import numpy as np

# --- Create a 3D array of shape (2, 3, 4) ---
arr = np.arange(1, 25, dtype=np.float64).reshape(2, 3, 4)

print("Array:\n", arr, "\n")

# --- Every attribute explained ---
print("shape    :", arr.shape)       # (2 blocks, 3 rows, 4 cols)
print("ndim     :", arr.ndim)        # number of dimensions = 3
print("size     :", arr.size)        # total elements = 2×3×4 = 24
print("dtype    :", arr.dtype)       # float64 → each element is 64-bit
print("itemsize :", arr.itemsize)    # bytes per single element (float64 = 8 bytes)
print("nbytes   :", arr.nbytes)      # total bytes = size × itemsize
print("strides  :", arr.strides)     # bytes to step to reach next element along each axis

# --- Manual verification: nbytes = size × itemsize ---
manual_nbytes = arr.size * arr.itemsize
print("\nManual nbytes (size × itemsize):", manual_nbytes)
print("Matches arr.nbytes?            ", manual_nbytes == arr.nbytes)

# --- Strides explained visually ---
print("\nStrides breakdown:")
print(f"  Along axis 0 (block) : {arr.strides[0]} bytes → jump over 1 block (3×4×8 = {3*4*8})")
print(f"  Along axis 1 (row)   : {arr.strides[1]} bytes → jump over 1 row  (4×8   = {4*8})")
print(f"  Along axis 2 (col)   : {arr.strides[2]} bytes → jump over 1 col  (1×8   = {1*8})")

Array:
 [[[ 1.  2.  3.  4.]
  [ 5.  6.  7.  8.]
  [ 9. 10. 11. 12.]]

 [[13. 14. 15. 16.]
  [17. 18. 19. 20.]
  [21. 22. 23. 24.]]] 

shape    : (2, 3, 4)
ndim     : 3
size     : 24
dtype    : float64
itemsize : 8
nbytes   : 192
strides  : (96, 32, 8)

Manual nbytes (size × itemsize): 192
Matches arr.nbytes?             True

Strides breakdown:
  Along axis 0 (block) : 96 bytes → jump over 1 block (3×4×8 = 96)
  Along axis 1 (row)   : 32 bytes → jump over 1 row  (4×8   = 32)
  Along axis 2 (col)   : 8 bytes → jump over 1 col  (1×8   = 8)


### How it works

- `shape (2, 3, 4)` — 2 blocks, each with 3 rows and 4 columns.
- `ndim 3` — three axes: block, row, column.
- `size 24` — total elements: 2 × 3 × 4 = 24.
- `dtype float64` — each element is a 64-bit floating point number.
- `itemsize 8` — one `float64` occupies 8 bytes in memory.
- `nbytes 192` — total memory used: 24 elements × 8 bytes = 192 bytes.
- `strides` — the number of bytes to skip in memory to move one step along each axis: moving across a column costs 8 bytes, across a row costs 32 bytes (4 cols × 8), across a block costs 96 bytes (3 rows × 4 cols × 8).

**Key takeaway:** `strides` reveals how NumPy navigates raw memory to present data as a multi-dimensional array — understanding strides explains why reshaping is instant (just changes strides, no data moved).

### Q16. `np.fromfunction`
Use `np.fromfunction` to create a 5×5 matrix where each element `[i][j]` equals `i * j` (the multiplication table starting from 0). Then create a 4×4 matrix where each element equals `i + j`.

In [19]:
# Q16: Build arrays from index-based formulas using np.fromfunction

import numpy as np

# --- 5×5 multiplication table: element [i][j] = i * j ---
mult_table = np.fromfunction(lambda i, j: i * j, shape=(5, 5), dtype=int)

# --- 4×4 sum table: element [i][j] = i + j ---
sum_table = np.fromfunction(lambda i, j: i + j, shape=(4, 4), dtype=int)

print("5×5 Multiplication table (i * j):\n", mult_table, "\n")
print("4×4 Sum table (i + j):\n", sum_table)

5×5 Multiplication table (i * j):
 [[ 0  0  0  0  0]
 [ 0  1  2  3  4]
 [ 0  2  4  6  8]
 [ 0  3  6  9 12]
 [ 0  4  8 12 16]] 

4×4 Sum table (i + j):
 [[0 1 2 3]
 [1 2 3 4]
 [2 3 4 5]
 [3 4 5 6]]


### How it works

- `np.fromfunction(func, shape)` calls `func(i, j)` where `i` and `j` are full index arrays — not single values — so the formula is applied to all positions at once (vectorized).
- Think of it as: NumPy passes a grid of row indices and a grid of column indices, then your lambda computes the result for every cell simultaneously.
- `dtype=int` makes the index arrays integers — without it they default to `float64` and results like `0 * 1` would show as `0.0`.
- The multiplication table row 0 is all zeros (0 × anything = 0), and the diagonal is perfect squares (i × i).
- You can use any mathematical formula involving `i` and `j` — making `fromfunction` ideal for building structured matrices without loops.

**Key takeaway:** `np.fromfunction` lets you define an entire matrix by a formula over its indices — no loops needed, and NumPy applies it to every cell in one vectorized pass.

### Q17. `np.fromiter`
Use `np.fromiter` to create a NumPy array from a generator expression that yields the first 20 Fibonacci numbers. Specify `count=20` and `dtype=int`.

In [20]:
# Q17: Build a NumPy array from a generator using np.fromiter

import numpy as np

# --- Fibonacci generator (yields values one at a time, never stores the full sequence) ---
def fibonacci():
    a, b = 0, 1
    while True:          # infinite generator — np.fromiter stops it at count=20
        yield a
        a, b = b, a + b

# --- Convert generator output into a NumPy array ---
fib_arr = np.fromiter(fibonacci(), dtype=int, count=20)

print("First 20 Fibonacci numbers:\n", fib_arr)
print("\nshape :", fib_arr.shape)
print("dtype :", fib_arr.dtype)

First 20 Fibonacci numbers:
 [   0    1    1    2    3    5    8   13   21   34   55   89  144  233
  377  610  987 1597 2584 4181]

shape : (20,)
dtype : int64


### How it works

- A **generator** (function using `yield`) produces values one at a time without storing them all in memory — it is like a tap that drips numbers on demand.
- `np.fromiter(generator, dtype, count)` pulls exactly `count` values from the generator and stores them in a compact NumPy array.
- `count=20` is important: without it, NumPy cannot know how many elements are coming and must resize the array repeatedly — specifying `count` pre-allocates memory in one shot.
- The `while True` loop makes the generator infinite — `np.fromiter` stops pulling after 20 values, so there is no infinite loop risk.
- `fromiter` only works with **1D arrays** — for 2D output you would need to reshape after.

**Key takeaway:** Use `np.fromiter` to efficiently convert any Python generator into a NumPy array — always specify `count` when known so memory is pre-allocated in one step.

### Q18. Structured Arrays
Create a structured NumPy array with fields `name` (string, max 20 chars), `age` (int), and `salary` (float). Add at least 4 records. Access individual fields by name (e.g., `data['name']`). Demonstrate that structured arrays are the NumPy equivalent of simple database tables.

In [21]:
# Q18: Create and query a structured NumPy array (like a simple database table)

import numpy as np

# --- Define the structure: each field has a name and dtype ---
dtype = np.dtype([
    ('name',   'U20'),       # Unicode string, max 20 characters
    ('age',    np.int32),    # 32-bit integer
    ('salary', np.float64)   # 64-bit float
])

# --- Create the structured array with 4 records ---
data = np.array([
    ('Alice',   31, 85000.0),
    ('Bob',     27, 72000.0),
    ('Chetan',  26, 91500.0),
    ('Diana',   35, 110000.0)
], dtype=dtype)

# --- Access entire array ---
print("Full table:\n", data, "\n")

# --- Access individual fields (like SQL column SELECT) ---
print("Names:   ", data['name'])
print("Ages:    ", data['age'])
print("Salaries:", data['salary'], "\n")

# --- Access a single record (like a table row) ---
print("Record 0:", data[0])
print("  name  :", data[0]['name'])
print("  age   :", data[0]['age'])
print("  salary:", data[0]['salary'], "\n")

# --- Filter rows: employees with salary > 80000 (like SQL WHERE) ---
high_earners = data[data['salary'] > 80000]
print("Salary > 80,000:\n", high_earners, "\n")

# --- Sort by age (like SQL ORDER BY) ---
sorted_by_age = np.sort(data, order='age')
print("Sorted by age:\n", sorted_by_age)

Full table:
 [('Alice', 31,  85000.) ('Bob', 27,  72000.) ('Chetan', 26,  91500.)
 ('Diana', 35, 110000.)] 

Names:    ['Alice' 'Bob' 'Chetan' 'Diana']
Ages:     [31 27 26 35]
Salaries: [ 85000.  72000.  91500. 110000.] 

Record 0: ('Alice', 31, 85000.0)
  name  : Alice
  age   : 31
  salary: 85000.0 

Salary > 80,000:
 [('Alice', 31,  85000.) ('Chetan', 26,  91500.) ('Diana', 35, 110000.)] 

Sorted by age:
 [('Chetan', 26,  91500.) ('Bob', 27,  72000.) ('Alice', 31,  85000.)
 ('Diana', 35, 110000.)]


### How it works

- A structured array uses `np.dtype([('field', type), ...])` to define columns — each row holds one record with named fields of mixed types.
- `data['name']` extracts an entire column by name — just like `SELECT name FROM table` in SQL.
- `data[0]` retrieves a single row; `data[0]['salary']` drills into one cell.
- Boolean indexing (`data[data['salary'] > 80000]`) filters rows by a condition — equivalent to a SQL `WHERE` clause.
- `np.sort(data, order='field')` sorts all rows by a chosen field — equivalent to `ORDER BY`.

**Key takeaway:** Structured arrays give NumPy a table-like structure with named, typed columns — useful for lightweight record storage without needing pandas, though pandas `DataFrame` is more powerful for complex queries.

### Q19. `np.concatenate`, `np.vstack`, `np.hstack`
Create two 2D arrays of shape `(3, 4)`. Demonstrate:
- Concatenating along axis 0 (rows) with `np.concatenate`
- Concatenating along axis 1 (columns) with `np.concatenate`
- Doing the same with `np.vstack` and `np.hstack`
- Show the resulting shapes

In [22]:
# Q19: Combine arrays using concatenate, vstack, and hstack

import numpy as np

# --- Two source arrays of shape (3, 4) ---
a = np.arange(1,  13).reshape(3, 4)    # values  1–12
b = np.arange(13, 25).reshape(3, 4)    # values 13–24

print("Array a (3×4):\n", a, "\n")
print("Array b (3×4):\n", b, "\n")

# --- Concatenate along axis 0: stack rows → shape (6, 4) ---
concat_axis0 = np.concatenate([a, b], axis=0)
print("concatenate axis=0 (rows stacked) → shape", concat_axis0.shape, ":\n", concat_axis0, "\n")

# --- Concatenate along axis 1: stack columns → shape (3, 8) ---
concat_axis1 = np.concatenate([a, b], axis=1)
print("concatenate axis=1 (cols stacked) → shape", concat_axis1.shape, ":\n", concat_axis1, "\n")

# --- vstack: vertical stack (same as concatenate axis=0) → shape (6, 4) ---
vstacked = np.vstack([a, b])
print("vstack → shape", vstacked.shape, ":\n", vstacked, "\n")

# --- hstack: horizontal stack (same as concatenate axis=1) → shape (3, 8) ---
hstacked = np.hstack([a, b])
print("hstack → shape", hstacked.shape, ":\n", hstacked, "\n")

# --- Verify vstack and hstack match concatenate ---
print("vstack matches concatenate axis=0:", np.array_equal(vstacked, concat_axis0))
print("hstack matches concatenate axis=1:", np.array_equal(hstacked, concat_axis1))

Array a (3×4):
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]] 

Array b (3×4):
 [[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]] 

concatenate axis=0 (rows stacked) → shape (6, 4) :
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]] 

concatenate axis=1 (cols stacked) → shape (3, 8) :
 [[ 1  2  3  4 13 14 15 16]
 [ 5  6  7  8 17 18 19 20]
 [ 9 10 11 12 21 22 23 24]] 

vstack → shape (6, 4) :
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]] 

hstack → shape (3, 8) :
 [[ 1  2  3  4 13 14 15 16]
 [ 5  6  7  8 17 18 19 20]
 [ 9 10 11 12 21 22 23 24]] 

vstack matches concatenate axis=0: True
hstack matches concatenate axis=1: True


### How it works

- `np.concatenate([a, b], axis=0)` joins arrays **vertically** — rows are appended, so `(3,4) + (3,4) → (6,4)`. Column count must match.
- `np.concatenate([a, b], axis=1)` joins arrays **horizontally** — columns are appended, so `(3,4) + (3,4) → (3,8)`. Row count must match.
- `np.vstack` is a convenience shortcut for `axis=0` — "v" stands for vertical (stacking downward).
- `np.hstack` is a convenience shortcut for `axis=1` — "h" stands for horizontal (stacking sideways).
- A simple rule: `axis=0` grows the **row** dimension; `axis=1` grows the **column** dimension.

**Key takeaway:** `vstack` and `hstack` are readable shortcuts for the two most common concatenation directions — under the hood they call `concatenate` with `axis=0` and `axis=1` respectively.

### Q20. `np.split`, `np.hsplit`, `np.vsplit`
Take a 6×6 array and demonstrate:
- `np.split` into 3 equal parts along axis 0
- `np.hsplit` into 2 equal halves
- `np.vsplit` into 3 equal parts
- `np.split` with explicit split indices `[2, 4]` (unequal splits)

In [23]:
# Q20: Split arrays using np.split, np.hsplit, np.vsplit

import numpy as np

# --- 6×6 source array ---
arr = np.arange(1, 37).reshape(6, 6)
print("Source array (6×6):\n", arr, "\n")

# --- np.split into 3 equal parts along axis 0 (rows) → three (2×6) arrays ---
row_splits = np.split(arr, 3, axis=0)
print("np.split into 3 along axis=0 → each shape:", row_splits[0].shape)
for i, part in enumerate(row_splits):
    print(f"  Part {i}:\n", part)

# --- np.hsplit into 2 equal halves (splits columns) → two (6×3) arrays ---
h_splits = np.hsplit(arr, 2)
print("\nnp.hsplit into 2 → each shape:", h_splits[0].shape)
for i, part in enumerate(h_splits):
    print(f"  Part {i}:\n", part)

# --- np.vsplit into 3 equal parts (splits rows) → three (2×6) arrays ---
v_splits = np.vsplit(arr, 3)
print("\nnp.vsplit into 3 → each shape:", v_splits[0].shape)
for i, part in enumerate(v_splits):
    print(f"  Part {i}:\n", part)

# --- np.split with explicit indices [2, 4]: rows 0–1, rows 2–3, rows 4–5 ---
unequal = np.split(arr, [2, 4], axis=0)   # cut before row 2 and before row 4
print("\nnp.split with indices [2,4] along axis=0:")
for i, part in enumerate(unequal):
    print(f"  Part {i} shape {part.shape}:\n", part)

Source array (6×6):
 [[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]
 [25 26 27 28 29 30]
 [31 32 33 34 35 36]] 

np.split into 3 along axis=0 → each shape: (2, 6)
  Part 0:
 [[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]
  Part 1:
 [[13 14 15 16 17 18]
 [19 20 21 22 23 24]]
  Part 2:
 [[25 26 27 28 29 30]
 [31 32 33 34 35 36]]

np.hsplit into 2 → each shape: (6, 3)
  Part 0:
 [[ 1  2  3]
 [ 7  8  9]
 [13 14 15]
 [19 20 21]
 [25 26 27]
 [31 32 33]]
  Part 1:
 [[ 4  5  6]
 [10 11 12]
 [16 17 18]
 [22 23 24]
 [28 29 30]
 [34 35 36]]

np.vsplit into 3 → each shape: (2, 6)
  Part 0:
 [[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]
  Part 1:
 [[13 14 15 16 17 18]
 [19 20 21 22 23 24]]
  Part 2:
 [[25 26 27 28 29 30]
 [31 32 33 34 35 36]]

np.split with indices [2,4] along axis=0:
  Part 0 shape (2, 6):
 [[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]
  Part 1 shape (2, 6):
 [[13 14 15 16 17 18]
 [19 20 21 22 23 24]]
  Part 2 shape (2, 6):
 [[25 26 27 28 29 30]
 [31 32 

### How it works

- `np.split(arr, N, axis)` divides the array into `N` equal parts along the given axis — raises an error if the axis size is not divisible by `N`.
- `np.vsplit(arr, N)` is a shortcut for `np.split(arr, N, axis=0)` — splits **rows** into N equal groups.
- `np.hsplit(arr, N)` is a shortcut for `np.split(arr, N, axis=1)` — splits **columns** into N equal groups.
- Passing a list of indices like `[2, 4]` to `np.split` creates **cut points**: part 0 gets rows before index 2, part 1 gets rows 2–3, part 2 gets rows 4 onward — allowing unequal splits.
- Split is the exact inverse of `vstack`/`hstack` — what one joins, the other can break apart.

**Key takeaway:** Use `vsplit`/`hsplit` for equal splits along rows or columns, and pass an index list to `np.split` when you need unequal cuts at specific positions.


### Q21. `np.tile` and `np.repeat`
Demonstrate the difference between `np.tile` and `np.repeat` using the array `[1, 2, 3]`:
- `np.tile([1, 2, 3], 4)` — repeat the whole pattern
- `np.repeat([1, 2, 3], 4)` — repeat each element
- Then apply both to a 2D array and show how the `axis` parameter changes behavior

In [24]:
# Q21: Understand the difference between np.tile and np.repeat

import numpy as np

base = np.array([1, 2, 3])

# --- 1D: tile repeats the whole pattern ---
tiled = np.tile(base, 4)            # [1,2,3] repeated 4 times as a block
print("np.tile([1,2,3], 4):  ", tiled)

# --- 1D: repeat repeats each element ---
repeated = np.repeat(base, 4)       # each element repeated 4 times in place
print("np.repeat([1,2,3], 4):", repeated, "\n")

# --- 2D source array ---
arr2d = np.array([[1, 2],
                  [3, 4]])
print("2D source array:\n", arr2d, "\n")

# --- np.tile on 2D: (2,3) means tile 2× along rows, 3× along cols ---
tiled_2d = np.tile(arr2d, (2, 3))
print("np.tile(arr2d, (2,3)) → shape", tiled_2d.shape, ":\n", tiled_2d, "\n")

# --- np.repeat on 2D along axis=0: repeat each ROW 3 times ---
repeat_ax0 = np.repeat(arr2d, 3, axis=0)
print("np.repeat(arr2d, 3, axis=0) → shape", repeat_ax0.shape, ":\n", repeat_ax0, "\n")

# --- np.repeat on 2D along axis=1: repeat each COLUMN element 3 times ---
repeat_ax1 = np.repeat(arr2d, 3, axis=1)
print("np.repeat(arr2d, 3, axis=1) → shape", repeat_ax1.shape, ":\n", repeat_ax1)

np.tile([1,2,3], 4):   [1 2 3 1 2 3 1 2 3 1 2 3]
np.repeat([1,2,3], 4): [1 1 1 1 2 2 2 2 3 3 3 3] 

2D source array:
 [[1 2]
 [3 4]] 

np.tile(arr2d, (2,3)) → shape (4, 6) :
 [[1 2 1 2 1 2]
 [3 4 3 4 3 4]
 [1 2 1 2 1 2]
 [3 4 3 4 3 4]] 

np.repeat(arr2d, 3, axis=0) → shape (6, 2) :
 [[1 2]
 [1 2]
 [1 2]
 [3 4]
 [3 4]
 [3 4]] 

np.repeat(arr2d, 3, axis=1) → shape (2, 6) :
 [[1 1 1 2 2 2]
 [3 3 3 4 4 4]]


### How it works

- `np.tile(arr, N)` repeats the **entire array** as a block `N` times — like copy-pasting the pattern end to end: `[1,2,3,1,2,3,1,2,3,1,2,3]`.
- `np.repeat(arr, N)` repeats **each element** `N` times in place: `[1,1,1,1,2,2,2,2,3,3,3,3]`.
- For 2D, `np.tile(arr, (rows, cols))` tiles the whole block in both directions simultaneously.
- `np.repeat` with `axis=0` stretches each **row** downward; with `axis=1` it stretches each **column** element rightward.
- Think of `tile` as "photocopying the page", and `repeat` as "stretching each cell of the table".

**Key takeaway:** `tile` repeats the whole pattern as a block; `repeat` expands individual elements — choosing the right one depends on whether you want pattern repetition or element-level stretching.


### Q22. Creating Arrays from Files
Write a 2D NumPy array to a text file using `np.savetxt` (with a custom delimiter and format). Read it back using `np.loadtxt`. Verify the round-trip by comparing original and loaded arrays with `np.array_equal`. Also save/load in binary `.npy` format using `np.save`/`np.load`.

In [25]:
# Q22: Save and load NumPy arrays — text (.txt) and binary (.npy) formats

import numpy as np

# --- Source array ---
original = np.array([[1.5, 2.7, 3.1],
                     [4.0, 5.9, 6.3],
                     [7.2, 8.8, 9.0]])

# ── TEXT FORMAT (.txt) ───────────────────────────────────────────────

# Save: comma-delimited, 2 decimal places, with a header comment
np.savetxt("array_data.txt",
           original,
           delimiter=",",        # column separator
           fmt="%.2f",           # format each value to 2 decimal places
           header="col1,col2,col3")

# Load back: must specify same delimiter; skip header row automatically
loaded_txt = np.loadtxt("array_data.txt", delimiter=",")

print("Original array:\n", original, "\n")
print("Loaded from .txt:\n", loaded_txt, "\n")
print("Round-trip match (.txt):", np.array_equal(original, loaded_txt), "\n")

# Peek at the raw file to show what was written
print("Raw .txt file contents:")
with open("array_data.txt") as f:
    print(f.read())

# ── BINARY FORMAT (.npy) ─────────────────────────────────────────────

# Save: preserves dtype, shape, and precision exactly — no formatting needed
np.save("array_data.npy", original)

# Load: restores array exactly as saved
loaded_npy = np.load("array_data.npy")

print("Loaded from .npy:\n", loaded_npy, "\n")
print("Round-trip match (.npy):", np.array_equal(original, loaded_npy))
print("dtype preserved:", loaded_npy.dtype)

Original array:
 [[1.5 2.7 3.1]
 [4.  5.9 6.3]
 [7.2 8.8 9. ]] 

Loaded from .txt:
 [[1.5 2.7 3.1]
 [4.  5.9 6.3]
 [7.2 8.8 9. ]] 

Round-trip match (.txt): True 

Raw .txt file contents:
# col1,col2,col3
1.50,2.70,3.10
4.00,5.90,6.30
7.20,8.80,9.00

Loaded from .npy:
 [[1.5 2.7 3.1]
 [4.  5.9 6.3]
 [7.2 8.8 9. ]] 

Round-trip match (.npy): True
dtype preserved: float64


### How it works

- `np.savetxt` writes array values as human-readable text; `delimiter` controls the separator and `fmt` controls decimal precision — but formatting can cause tiny rounding differences.
- `np.loadtxt` reads the file back; it automatically skips lines starting with `#` (the header), and the `delimiter` must match what was used when saving.
- `np.array_equal` returns `True` only if every element matches exactly — a good sanity check for round-trips.
- `np.save` writes a compact binary `.npy` file that stores the exact dtype, shape, and all bytes — no precision loss, no formatting needed.
- `np.load` restores the binary file perfectly — faster to read and write than text, and safer for float precision.

**Key takeaway:** Use `.txt` when humans need to read or share the file; use `.npy` when you need speed, exact precision, and dtype preservation — binary is always the safer choice for numerical data.

### Q23. `np.meshgrid`
Use `np.meshgrid` to create coordinate grids for x values `[0, 1, 2, 3]` and y values `[0, 1, 2]`. Print both grid arrays and explain their shapes. Then compute `z = x**2 + y**2` over the whole grid in a single expression (no loops) — this is the foundation of 3D surface plotting.

In [26]:
# Q23: Build coordinate grids with np.meshgrid and compute a surface formula

import numpy as np

# --- Define axis values ---
x_vals = np.array([0, 1, 2, 3])    # 4 points along x-axis
y_vals = np.array([0, 1, 2])       # 3 points along y-axis

# --- Create 2D coordinate grids ---
X, Y = np.meshgrid(x_vals, y_vals)
# X repeats x_vals across each row → shape (3, 4)
# Y repeats y_vals down each col  → shape (3, 4)

print("x_vals:", x_vals)
print("y_vals:", y_vals, "\n")

print("X grid (x coordinate at every point):\n", X)
print("Shape:", X.shape, "← (len(y_vals), len(x_vals))\n")

print("Y grid (y coordinate at every point):\n", Y)
print("Shape:", Y.shape, "\n")

# --- Compute z = x² + y² over the entire grid (no loops) ---
Z = X**2 + Y**2
print("Z = X² + Y²:\n", Z)
print("Shape:", Z.shape, "\n")

# --- Show what each cell means ---
print("Verification — point (x=2, y=1): Z =", 2**2 + 1**2, "| grid says:", Z[1, 2])
print("Verification — point (x=3, y=2): Z =", 3**2 + 2**2, "| grid says:", Z[2, 3])

x_vals: [0 1 2 3]
y_vals: [0 1 2] 

X grid (x coordinate at every point):
 [[0 1 2 3]
 [0 1 2 3]
 [0 1 2 3]]
Shape: (3, 4) ← (len(y_vals), len(x_vals))

Y grid (y coordinate at every point):
 [[0 0 0 0]
 [1 1 1 1]
 [2 2 2 2]]
Shape: (3, 4) 

Z = X² + Y²:
 [[ 0  1  4  9]
 [ 1  2  5 10]
 [ 4  5  8 13]]
Shape: (3, 4) 

Verification — point (x=2, y=1): Z = 5 | grid says: 5
Verification — point (x=3, y=2): Z = 13 | grid says: 13


### How it works

- `np.meshgrid(x_vals, y_vals)` produces two 2D arrays `X` and `Y`, each of shape `(len(y), len(x))` — every cell holds the x or y coordinate of that grid point.
- `X` repeats `x_vals` identically across every row; `Y` repeats `y_vals` identically down every column — together they describe every (x, y) combination on a grid.
- Once you have `X` and `Y`, any formula like `Z = X**2 + Y**2` computes the result at all grid points simultaneously — no nested loops needed.
- `Z[row, col]` corresponds to the point `(x_vals[col], y_vals[row])` — rows index y, columns index x.
- This is the foundation of 3D surface plots: `X`, `Y`, and `Z` are the three coordinate arrays passed to plotting libraries like Matplotlib's `plot_surface`.

**Key takeaway:** `np.meshgrid` converts two 1D axis arrays into full 2D coordinate grids, enabling vectorized evaluation of any formula over every grid point — the core pattern for surface and contour plotting.

### Q24. Array Comparison and `np.array_equal`
Create two arrays that look equal but differ in dtype. Use `==` (element-wise), `np.array_equal`, and `np.allclose` to compare them. Explain the difference between all three comparison methods and when you'd use each.

In [28]:
# Q24: Compare arrays using ==, np.array_equal, and np.allclose

import numpy as np

# --- Two arrays that look equal but differ in dtype ---
a = np.array([1.0, 2.0, 3.0], dtype=np.float64)
b = np.array([1,   2,   3  ], dtype=np.int32)

# --- Also create a float array with tiny rounding differences ---
c = np.array([1.0000000001, 2.0000000001, 3.0000000001], dtype=np.float64)

print("a (float64):", a, "| dtype:", a.dtype)
print("b (int32)  :", b, "| dtype:", b.dtype)
print("c (float64):", c, "| dtype:", c.dtype, "\n")

# ── METHOD 1: == (element-wise comparison) ───────────────────────────
elementwise = a == b           # returns a bool array, ignores dtype
print("a == b (element-wise):", elementwise)
print("All True?            :", elementwise.all(), "\n")

# ── METHOD 2: np.array_equal ─────────────────────────────────────────
equal_ab = np.array_equal(a, b)    # True only if same shape, dtype, AND values
equal_ac = np.array_equal(a, c)    # False — tiny float differences count
print("np.array_equal(a, b):", equal_ab, "← dtype differs, but values match")
print("np.array_equal(a, c):", equal_ac, "← tiny float diff → not equal\n")

# ── METHOD 3: np.allclose ────────────────────────────────────────────
close_ab = np.allclose(a, b)           # tolerates dtype difference
close_ac = np.allclose(a, c)           # tolerates tiny float differences
close_ac_strict = np.allclose(a, c, rtol=0, atol=1e-12)   # tighter tolerance → False
print("np.allclose(a, b)              :", close_ab, "← values close enough")
print("np.allclose(a, c)              :", close_ac, "← tiny diff within default tolerance")
print("np.allclose(a, c, atol=1e-12) :", close_ac_strict, "← strict tolerance catches diff\n")

# ── Summary table ────────────────────────────────────────────────────
print("─" * 52)
print(f"{'Method':<28} {'a vs b':<12} {'a vs c'}")
print("─" * 52)
print(f"{'== (element-wise).all()':<28} {str((a==b).all()):<12} {str((a==c).all())}")
print(f"{'np.array_equal':<28} {str(equal_ab):<12} {str(equal_ac)}")
print(f"{'np.allclose (default)':<28} {str(close_ab):<12} {str(close_ac)}")
print("─" * 52)

a (float64): [1. 2. 3.] | dtype: float64
b (int32)  : [1 2 3] | dtype: int32
c (float64): [1. 2. 3.] | dtype: float64 

a == b (element-wise): [ True  True  True]
All True?            : True 

np.array_equal(a, b): True ← dtype differs, but values match
np.array_equal(a, c): False ← tiny float diff → not equal

np.allclose(a, b)              : True ← values close enough
np.allclose(a, c)              : True ← tiny diff within default tolerance
np.allclose(a, c, atol=1e-12) : False ← strict tolerance catches diff

────────────────────────────────────────────────────
Method                       a vs b       a vs c
────────────────────────────────────────────────────
== (element-wise).all()      True         False
np.array_equal               True         False
np.allclose (default)        True         True
────────────────────────────────────────────────────


### How it works

- `==` returns a **bool array** — one `True`/`False` per element pair. It checks values only, ignores dtype, and is useful when you need to know *which* elements match, not just *whether* all match.
- `np.array_equal(a, b)` returns a single `True` or `False` — it requires identical shape and values but is surprisingly dtype-lenient: `float64(1.0)` and `int32(1)` are considered equal.
- `np.allclose(a, b)` returns a single `True` or `False` and **tolerates small floating-point differences** using `rtol` (relative) and `atol` (absolute) tolerance thresholds — essential when comparing results of different floating-point calculations.
- Default `allclose` tolerance is `rtol=1e-5, atol=1e-8` — enough to absorb normal float arithmetic drift, but tighten with `atol=1e-12` for stricter checks.
- Use `==` for masking/filtering, `array_equal` for exact round-trip checks (e.g. after save/load), and `allclose` whenever floats went through math operations.

**Key takeaway:** Never use `==` alone to decide if two float arrays are "the same" — use `np.allclose` instead, because floating-point arithmetic almost never produces bit-identical results.

### Q25. Challenge — Building a Mini Array Factory
Write a function `make_arrays(spec)` that accepts a list of specification dictionaries and returns a list of NumPy arrays. Each spec dict can have keys:
- `"type"`: one of `"zeros"`, `"ones"`, `"eye"`, `"random"`, `"linspace"`, `"arange"`
- `"shape"`: tuple (for zeros/ones/random)
- `"dtype"`: string dtype name
- `"start"`, `"stop"`, `"num"` or `"step"` (for linspace/arange)
Handle unknown types with a custom `UnknownArrayTypeError`. Test with at least 6 different spec dicts, printing each resulting array's shape and dtype.

In [29]:
# Q25: Mini array factory — create NumPy arrays from specification dictionaries

import numpy as np

# --- Custom exception for unknown array types ---
class UnknownArrayTypeError(Exception):
    pass

# --- Factory function ---
def make_arrays(spec):
    """Accept a list of spec dicts, return a list of NumPy arrays."""
    results = []

    for s in spec:
        arr_type = s.get("type")
        dtype    = s.get("dtype", None)    # optional for all types

        if arr_type == "zeros":
            arr = np.zeros(s["shape"], dtype=dtype)

        elif arr_type == "ones":
            arr = np.ones(s["shape"], dtype=dtype)

        elif arr_type == "eye":
            # shape[0] gives the N for an N×N identity matrix
            arr = np.eye(s["shape"][0], dtype=dtype)

        elif arr_type == "random":
            arr = np.random.rand(*s["shape"])           # always float64
            if dtype:
                arr = arr.astype(dtype)

        elif arr_type == "linspace":
            arr = np.linspace(s["start"], s["stop"], s["num"], dtype=dtype)

        elif arr_type == "arange":
            arr = np.arange(s["start"], s["stop"], s.get("step", 1), dtype=dtype)

        else:
            raise UnknownArrayTypeError(
                f"Unknown array type '{arr_type}'. "
                f"Valid types: zeros, ones, eye, random, linspace, arange."
            )

        results.append(arr)

    return results

# --- Test specs: 6 different configurations ---
specs = [
    {"type": "zeros",    "shape": (3, 4),  "dtype": "float32"},
    {"type": "ones",     "shape": (2, 5),  "dtype": "int16"},
    {"type": "eye",      "shape": (4, 4),  "dtype": "float64"},
    {"type": "random",   "shape": (3, 3),  "dtype": "float32"},
    {"type": "linspace", "start": 0, "stop": 1,   "num": 6,  "dtype": "float64"},
    {"type": "arange",   "start": 0, "stop": 20,  "step": 3, "dtype": "int32"},
]

arrays = make_arrays(specs)

print("=" * 48)
for spec, arr in zip(specs, arrays):
    print(f"type='{spec['type']}'")
    print(f"  shape : {arr.shape}")
    print(f"  dtype : {arr.dtype}")
    print(f"  array : {arr if arr.ndim == 1 else ''}")
    if arr.ndim > 1:
        print(arr)
    print()

# --- Test custom exception ---
print("=" * 48)
print("Testing UnknownArrayTypeError:")
try:
    make_arrays([{"type": "diagonal", "shape": (3, 3)}])
except UnknownArrayTypeError as e:
    print(f"  Caught → {e}")

type='zeros'
  shape : (3, 4)
  dtype : float32
  array : 
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]

type='ones'
  shape : (2, 5)
  dtype : int16
  array : 
[[1 1 1 1 1]
 [1 1 1 1 1]]

type='eye'
  shape : (4, 4)
  dtype : float64
  array : 
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]

type='random'
  shape : (3, 3)
  dtype : float32
  array : 
[[0.5986585  0.15601864 0.15599452]
 [0.05808361 0.8661761  0.601115  ]
 [0.7080726  0.02058449 0.96990985]]

type='linspace'
  shape : (6,)
  dtype : float64
  array : [0.  0.2 0.4 0.6 0.8 1. ]

type='arange'
  shape : (7,)
  dtype : int32
  array : [ 0  3  6  9 12 15 18]

Testing UnknownArrayTypeError:
  Caught → Unknown array type 'diagonal'. Valid types: zeros, ones, eye, random, linspace, arange.


### How it works

- `make_arrays` loops over a list of dicts, reads the `"type"` key, and dispatches to the correct NumPy constructor — a classic **factory pattern**.
- `s.get("dtype", None)` safely returns `None` when dtype is omitted, letting NumPy use its own default.
- `*s["shape"]` unpacks the shape tuple as positional arguments — `np.random.rand` requires separate `(rows, cols)` arguments, unlike `np.zeros` which takes a tuple.
- `s.get("step", 1)` provides a safe default step of `1` for `arange` when the spec omits it.
- A custom exception class (`UnknownArrayTypeError`) inherits from `Exception` — raising it with a clear message makes debugging far easier than a silent failure or generic `KeyError`.

**Key takeaway:** The factory pattern — dispatching to different constructors based on a type string — is a fundamental design pattern in data pipelines, letting you define array configurations as data (dicts) rather than hardcoded function calls.